File: leet_code/cpp/39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */
#include "common.h"
// @lc code=start
class Solution {
 public:
  vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    vector<vector<int>> res;
    sort(candidates.begin(), candidates.end());
    vector<int> tmp_res;
    search(candidates, 0, tmp_res, target, res);
    return res;
  }

 private:
  void search(vector<int>& candidates, int next_pos, vector<int>& tmp_res,
              int target, vector<vector<int>>& res) {
    if (target == 0) {
      res.push_back(tmp_res);
    }
    if (next_pos == candidates.size() || target - candidates[next_pos] < 0) {
      return;
    }
    tmp_res.push_back(candidates[next_pos]);
    search(candidates, next_pos, tmp_res, target - candidates[next_pos], res);
    tmp_res.pop_back();
    search(candidates, next_pos + 1, tmp_res, target, res);
  }
};
// @lc code=end

File: python/39.组合总和.py
#
# @lc app=leetcode.cn id=39 lang=python3
#
# [39] 组合总和
#
import bisect
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        self.biSearch(candidates, target, [], res)
        return res

    def search(self, candidates, target, path, res):
        if target < 0:
            return
        if target == 0:
            res.append(path)
            return
        for idx,item in enumerate(candidates):
            self.search(candidates[idx:], target-item, path+[item], res)

    def biSearch(self, candidates, target, path, res):
        if not candidates or target < 0:
            return
        if target == 0:
            res.append(path)
            return
        id_s = bisect.bisect_left(candidates, target)

        for idx,item in enumerate(candidates[:id_s+1]):
            self.biSearch(candidates[idx:id_s+1], target-item, path+[item], res)

if __name__ == "__main__":
    s = Solution()
    res = s.combinationSum([2,3,5], 8)
    print(res)
